activesupport/lib/active_support/cache.rb

```diff
@@ -312,28 +313,31 @@ def mute
       #   end
       #   cache.fetch('foo') # => "bar"
       def fetch(name, options = nil)
+        master_key = Key.call(name)
```

Comment:
We have a few forms of keys floating around this method and the naming is piling up. How would you communicate the role of the "master" key here? How does it differ from the "normalized" key, or the name?

Seems we have key components (name), normalized key (serialized string key), and a Key instance (represents key components and serializes to a string).

In that case, 
- are we changing instrumentation behavior by passing the key instance rather than its components?
- should the Key instance be responsible for namespacing and "normalization" as well? Where's the boundary with cache store implementation concerns, like transforming the serialized keys?

Reply:
I'm still working my way through your comments. I wanted to go ahead and mention that I looked into this for a few hours:

> should the Key instance be responsible for namespacing and "normalization" as well? Where's the boundary with cache store implementation concerns, like transforming the serialized keys?

I was able to push the responsibility to the Key class, but we're not able to eliminate calls to the `normalize_key` method calls, since other cache stores such as the redis store, extend this method https://github.com/rails/rails/blob/5431e17733366da1fd10f2cd3039d66a56012683/activesupport/lib/active_support/cache/redis_cache_store.rb#L434-L436

Since we can't get rid of that method call, the read/write/fetch method calls end up looking similar, however the Key class becomes more convoluted. I think it was worthwile to look into and I had wondered if it was possible before. Right now it doesn't look like the benefits outweight the costs.

> are we changing instrumentation behavior by passing the key instance rather than its components?

We are not changing the instrumetation behavior. In the `instrument` method, it calls `normalize_key` on the argument that is passed. If you don't pass in a key object, one gets created. The resulting generated strings will be identical.

> We have a few forms of keys floating around this method and the naming is piling up. How would you communicate the role of the "master" key here? How does it differ from the "normalized" key, or the name?

I can do a better job of naming inside of the `fetch` method. This variable makes less sense now that the class is no longer "key master". 

I'm not ready for another review yet, but wanted to post my progress while it's fresh.